hashedindex/textparser.py

```diff
@@ -92,13 +96,14 @@ def word_tokenize(text, stopwords=_stopwords, ngrams=None, min_length=0, ignore_
     validate_stemmer(stemmer)
 
     text = re.sub(re.compile('\'s'), '', text)  # Simple heuristic
-    text = re.sub(_re_punctuation, '', text)
+    text = text if retain_punctuation else re.sub(_re_punctuation, '', text)
     text = text if retain_casing else text.lower()
 
     matched_tokens = re.findall(_re_token, text)
     for tokens in get_ngrams(matched_tokens, ngrams):
         for i in range(len(tokens)):
-            tokens[i] = tokens[i].strip(punctuation)
+            if not retain_punctuation:
+                tokens[i] = tokens[i].strip(punctuation)
```

Comment:
@MichaelAquilina Could it make sense to re-use `_re_punctuation` for this step?  I'm wondering if there may be any cases where `string.punctuation` and `_punctuation` may end up differing slightly, leading to unusual behaviour.

Reply:
I suppose this is for *within-token* punctuation characters (i.e. `pre-heat` -> `preheat`) - that's certainly a slightly different situation to splitting token boundaries.